# Econ433 project one: summarize safegraph datasets

By Michael T Kissiedu

Professor: Dr. Jin

Due date: Friday Feb 12th, 2021

# Introduction

      This project helped with understanding Safegraph data. The packages 'pandas,' 'NumPy,' 'matplotlib.pyplot,' 
and 're' were imported. The packages helped summarize and cleaned the data. The markdown cells used were to describe python commands and outcome for 

question in the project. Lastly, the Phython commands were described to make this project reader-friendly. The project's goal of was to help understand the 

Safegraph data. 

In [1]:
# The command below imports Python packages I will be using to help me summarize and clean Safegraph data.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re

Matplotlib is building the font cache; this may take a moment.


In [2]:
# The command below is to display two digits in pandas dataframe float numbers.
pd.options.display.float_format = '{:.2f}'.format

# Part 1: read-in nationwide data and create "wide sample."

    Part one included the analysis of the four datasets. Then the four datasets were linked together vertically into one data.

Below, the path to dataset 0309_raw.csv was imported.

In [3]:
data_09 = pd.read_csv('0309_raw.csv')

The command below showed how many observations were in '0309_raw.csv. The command answered the first question 1a, "how many observations are in each of the four datasets?"

In [4]:
len(data_09)

3856933

Relating to qestion 1a, 3,856,933 observations were in this dataset.

Below, the path to the dataset 0316-raw.csv was imported.

In [5]:
data_16 = pd.read_csv('0316_raw.csv')

TThe command below showed how many observations were in '0316_raw.csv. The command answered the first question 1a, "how many observations are in each of the four datasets?"

In [6]:
len(data_16)

3742506

Relating to qestion 1a, 3,742,506 observations were in this dataset.

Below, the path to the dataset 0323_raw.csv was imported.

In [7]:
# Here a path to dataset 0323_raw.csv on my computer was written. Python imported dataset 0330_raw.csv. It will have the name data_30. 
# The code below the path to the dataset, shows what the first five observations in the dataset looks like.
# After analyzing what the dataset looks like, question one as it pertains to this data will be answered.

data_23 = pd.read_csv('0323_raw.csv')
#data_23.head()

The command below showed how many observations were in '0323_raw.csv. The command answered the first question 1a, "how many observations are in each of the four datasets?"

In [8]:
len(data_23)

3637602

Relating to question 1a, 3,742,506 observations were in the dataset.

Below, the path to the dataset 0330_raw.csv was imported.

In [9]:
# Here a path to dataset 0330_raw.csv on my computer was written. Python will be importing dataset 0330_raw.csv. It will have the name data_30. 
# The code below the path to the dataset, shows what the first five observations in the dataset looks like.
# After analyzing what the dataset looks like, question one as it pertains to this data will be answered.

data_30 = pd.read_csv('0330_raw.csv')
#data_30.head()

The command below showed how many observations were in '0330_raw.csv. The command answered the first question 1a, "how many observations are in each of the four datasets?"

In [10]:
len(data_30)

3594714

Relating to question 1a, 3,594,714 observations were in dataset 0330_raw.csv.

Question one also ask, how many observations are there after stacking the four datasets together? 

Below, the four datasets are linked together vertically. This allowed the question directly above this statement to be answered.

In [11]:
# Here, I link the four nationwide datasets vertically into one data.
data = pd.concat([data_09, data_16, data_23, data_30], ignore_index=True)

In [12]:
# data.shape will tell me the observations and columns.
data.shape
len(data)

14831755

Relating to the question about how many observations does the four datasets stacked together have, a total of 14,831,755 (almost 15 million) observations are in the dataset (data) after stacking the four datasets together.

Question also ask, what is the unit of observation in the stacked data?

The code below answered the question directly above this statement.

In [13]:
# The code below showed a table. The table displayed information that helped answer the question.
data.head()

safegraph_place_id region     date_range_start  \
0  sg:0013bd02b3564a9187a52f8e2799233e     TX  2020-03-09 05:00:00   
1  sg:001955fa1c994b4c8c877316a66dd986     AL  2020-03-09 05:00:00   
2  sg:001e39c6b18645a5950b13a278b242c3     NY  2020-03-09 04:00:00   
3  sg:0021f605c7904ed2a3f7bd28d6c40844     IL  2020-03-09 05:00:00   
4  sg:0023804ee36c4056886ef54e5a0cb773     FL  2020-03-09 04:00:00   

        date_range_end           visits_by_day  raw_visit_counts  \
0  2020-03-16 05:00:00  [13,15,15,14,16,19,13]            105.00   
1  2020-03-16 05:00:00  [53,66,44,59,45,59,50]            376.00   
2  2020-03-16 04:00:00   [14,10,12,12,10,13,7]             78.00   
3  2020-03-16 05:00:00         [0,1,0,0,0,0,0]              1.00   
4  2020-03-16 04:00:00      [9,17,16,9,13,2,0]             66.00   

   raw_visitor_counts  
0               27.00  
1              314.00  
2               68.00  
3                1.00  
4               40.00

Relating to the final request of question 1a , the unit of observation is the place-id by week.

Now the focus is on question 1b, which ask, after conditioning on the state of interest (Wyoming) from the stacked data, how many observations are there?

The state of interest that is focused on Wyoming.

In [14]:
# The command below conditions on the state of interest.
w = data[data['region'] == 'WY'].copy()
# Wyoming is the state of interest. The letter 'w' is a variable representing Wyoming.

The command below showed how many observations were in the state of interest. The command answered the first question 1b, "how many observations were in the Wyoming dataset (w)?

In [15]:
len(w)

34328

After conditioning on Wyoming, a total of 34,328 observations are in the dataset.

Here the focus is on question 1c. Question 1C ask how many variables are there in the “wide sample” (w)?

The code below display how many row and columns are the "wide data." The columns represent the variables.

In [16]:
# The second number tells how many variables are there.
w.shape

(34328, 7)

Relating to question 1C, 7 variables are in the "wide" (w) dataset.

Here the focus is on the question what is the type of each variable? This question is from question 1C.

In [17]:
# The command below show the type of variables in the dataset for Wyoming.
w.dtypes

safegraph_place_id     object
region                 object
date_range_start       object
date_range_end         object
visits_by_day          object
raw_visit_counts      float64
raw_visitor_counts    float64
dtype: object

# Question 1c answer: 

* safegraph_place_id is an object

* Regions is an object/string. 

* date_range_start is an object/string. 

* date _rangeend is an object/string. 

* Visits_by_day is an object/string. 

* raw_visit_count is a flot64/integer. 

* Raw_visitor_counts is a float64/integer. 

# Part 2: Clean your “wide sample” and reshape it into “long sample”

Now, the focus is on Question 2a. The quesion states what are the begining date and ending date in 'wide sample.'

In [18]:
# This code helps me clean my data.
# I convert the date in my "wide" sample into python date with this code.
w['date_range_start'] = w['date_range_start'].str[:10]
w['date_range_end'] = w['date_range_end'].str[:10]


In [19]:
# This code shows how to convert the date into the python type (datetime64)
w['date_range_start'] = pd.to_datetime(w['date_range_start'])

w['date_range_end'] = pd.to_datetime(w['date_range_end'])
w[['date_range_start', 'date_range_end']].dtypes

date_range_start    datetime64[ns]
date_range_end      datetime64[ns]
dtype: object

In [20]:
# This code show the date time
w.head()

safegraph_place_id region date_range_start  \
466   sg:0e5816c2c3074793ae5fc6f9c3e6bb7e     WY       2020-03-09   
552   sg:111ac71acaee4d10b00ef01388ae3e98     WY       2020-03-09   
1399  sg:2a17146a35e149dabce27332a52f8d1f     WY       2020-03-09   
2694  sg:5380bbb5685d4e6eb8427f4b1d418772     WY       2020-03-09   
2752  sg:55dbc9ae0c514561b30d27f7b7d74710     WY       2020-03-09   

     date_range_end         visits_by_day  raw_visit_counts  \
466      2020-03-16      [11,5,1,2,1,7,7]             34.00   
552      2020-03-16  [13,13,13,5,13,2,11]             70.00   
1399     2020-03-16      [10,4,4,6,4,0,0]             28.00   
2694     2020-03-16      [5,8,5,4,14,8,5]             49.00   
2752     2020-03-16       [0,1,0,0,2,0,0]              3.00   

      raw_visitor_counts  
466                19.00  
552                41.00  
1399               13.00  
2694               44.00  
2752                2.00

In [21]:
# The code below sorted the date in an ascending order so figuring out the beginning and end date is easier.
w.sort_values(by = 'date_range_start', ascending = True)

safegraph_place_id region date_range_start  \
466       sg:0e5816c2c3074793ae5fc6f9c3e6bb7e     WY       2020-03-09   
2595023   sg:6f017677dcdf477497da3ad648dea6d2     WY       2020-03-09   
2594480   sg:5c7b5c301c414779923d647dae61d193     WY       2020-03-09   
2594163   sg:51f1989ec0144ebc8a32c2cef41c9ed6     WY       2020-03-09   
2593641   sg:3fa59933d6904f79bc66e88295f5f29d     WY       2020-03-09   
...                                       ...    ...              ...   
12451893  sg:165293ea82ad4029ae69aff64e6f556c     WY       2020-03-30   
12451556  sg:0b7035fc0ff94e15919f20fe8a65da08     WY       2020-03-30   
12450232  sg:dace4194bcfe4500a5203c8ef6c289a4     WY       2020-03-30   
12456611  sg:c12ffcb7895346d68032c4ec01f234f4     WY       2020-03-30   
14831677  sg:fd3bfb53b0cc4ca7bfc051d1ce2bb2e5     WY       2020-03-30   

         date_range_end          visits_by_day  raw_visit_counts  \
466          2020-03-16       [11,5,1,2,1,7,7]             34.00   
2595023      2020-03-16  [11,10,16,17,7,15,10]             86.00   
2594480      2020-03-16      [11,5,5,8,7,7,11]             54.00   
2594163      2020-03-16        [3,0,1,0,1,2,2]              9.00   
2593641      2020-03-16       [6,17,5,6,7,0,0]             41.00   
...                 ...                    ...               ...   
12451893     2020-04-06        [0,0,2,0,2,1,0]              5.00   
12451556     2020-04-06        [1,0,0,0,0,1,0]              2.00   
12450232     2020-04-06       [10,3,4,6,6,4,1]             34.00   
12456611     2020-04-06        [1,0,0,0,0,0,0]              1.00   
14831677     2020-04-06        [7,6,6,4,3,8,4]             38.00   

          raw_visitor_counts  
466                    19.00  
2595023                43.00  
2594480                21.00  
2594163                 9.00  
2593641                34.00  
...                      ...  
12451893                4.00  
12451556                2.00  
12450232               32.00  
12456611                1.00  
14831677               35.00  

[34328 rows x 7 columns]

In [22]:
# Again sorting the date, but in descending order so figuring out the beginning and end date is easier.
w.sort_values(by = 'date_range_end', ascending = False)

safegraph_place_id region date_range_start  \
14831677  sg:fd3bfb53b0cc4ca7bfc051d1ce2bb2e5     WY       2020-03-30   
12456954  sg:cd8d1cda019747c9bfd6dbaa1d3b8a65     WY       2020-03-30   
12451556  sg:0b7035fc0ff94e15919f20fe8a65da08     WY       2020-03-30   
12451893  sg:165293ea82ad4029ae69aff64e6f556c     WY       2020-03-30   
12452175  sg:203452d6f0d2461b929daa08b6b4579a     WY       2020-03-30   
...                                       ...    ...              ...   
2593641   sg:3fa59933d6904f79bc66e88295f5f29d     WY       2020-03-09   
2594163   sg:51f1989ec0144ebc8a32c2cef41c9ed6     WY       2020-03-09   
2594480   sg:5c7b5c301c414779923d647dae61d193     WY       2020-03-09   
2595023   sg:6f017677dcdf477497da3ad648dea6d2     WY       2020-03-09   
466       sg:0e5816c2c3074793ae5fc6f9c3e6bb7e     WY       2020-03-09   

         date_range_end          visits_by_day  raw_visit_counts  \
14831677     2020-04-06        [7,6,6,4,3,8,4]             38.00   
12456954     2020-04-06        [1,0,0,0,0,0,0]              1.00   
12451556     2020-04-06        [1,0,0,0,0,1,0]              2.00   
12451893     2020-04-06        [0,0,2,0,2,1,0]              5.00   
12452175     2020-04-06        [0,1,0,0,0,2,0]              3.00   
...                 ...                    ...               ...   
2593641      2020-03-16       [6,17,5,6,7,0,0]             41.00   
2594163      2020-03-16        [3,0,1,0,1,2,2]              9.00   
2594480      2020-03-16      [11,5,5,8,7,7,11]             54.00   
2595023      2020-03-16  [11,10,16,17,7,15,10]             86.00   
466          2020-03-16       [11,5,1,2,1,7,7]             34.00   

          raw_visitor_counts  
14831677               35.00  
12456954                1.00  
12451556                2.00  
12451893                4.00  
12452175                2.00  
...                      ...  
2593641                34.00  
2594163                 9.00  
2594480                21.00  
2595023                43.00  
466                    19.00  

[34328 rows x 7 columns]

Relating to question 2a, the answer to it was the "wide" sample (w's) begining date is 03/09/2020 and ending date is 04/06/2020

Here, the focus is on question 2b. Question two begin by stating, "in the raw data, the variable visits_by_day is text only. It contains a string of 7 numbers separated by comma (e.g. “[3,2,4,0,5,2,1]”). Please split them into 7 new variables (dailyvisits1, …, dailyvisits7), each of which is an integer." Then the questin is, how many variables in the “wide sample are there now?

In [23]:
# This code below shows the head so we can see how visits by day are arranged.
# They have the brackets. So more codes will be written in order to strip the brackets away. Then splitting the visit by day in 7 new column will be accomplished.
w.head()

safegraph_place_id region date_range_start  \
466   sg:0e5816c2c3074793ae5fc6f9c3e6bb7e     WY       2020-03-09   
552   sg:111ac71acaee4d10b00ef01388ae3e98     WY       2020-03-09   
1399  sg:2a17146a35e149dabce27332a52f8d1f     WY       2020-03-09   
2694  sg:5380bbb5685d4e6eb8427f4b1d418772     WY       2020-03-09   
2752  sg:55dbc9ae0c514561b30d27f7b7d74710     WY       2020-03-09   

     date_range_end         visits_by_day  raw_visit_counts  \
466      2020-03-16      [11,5,1,2,1,7,7]             34.00   
552      2020-03-16  [13,13,13,5,13,2,11]             70.00   
1399     2020-03-16      [10,4,4,6,4,0,0]             28.00   
2694     2020-03-16      [5,8,5,4,14,8,5]             49.00   
2752     2020-03-16       [0,1,0,0,2,0,0]              3.00   

      raw_visitor_counts  
466                19.00  
552                41.00  
1399               13.00  
2694               44.00  
2752                2.00

In [24]:
# So the couple of codes below will strip the brackets and split the visit by day in 7 new column.
w['visits_by_day'] = w['visits_by_day'].str.strip()


In [25]:
w['visits_by_day'] = w['visits_by_day'].str.replace('\[|\]', '')

In [26]:
# This code shows the head. What is below shows all the brackets stripped from visit_by_day.
w.head()

safegraph_place_id region date_range_start  \
466   sg:0e5816c2c3074793ae5fc6f9c3e6bb7e     WY       2020-03-09   
552   sg:111ac71acaee4d10b00ef01388ae3e98     WY       2020-03-09   
1399  sg:2a17146a35e149dabce27332a52f8d1f     WY       2020-03-09   
2694  sg:5380bbb5685d4e6eb8427f4b1d418772     WY       2020-03-09   
2752  sg:55dbc9ae0c514561b30d27f7b7d74710     WY       2020-03-09   

     date_range_end       visits_by_day  raw_visit_counts  raw_visitor_counts  
466      2020-03-16      11,5,1,2,1,7,7             34.00               19.00  
552      2020-03-16  13,13,13,5,13,2,11             70.00               41.00  
1399     2020-03-16      10,4,4,6,4,0,0             28.00               13.00  
2694     2020-03-16      5,8,5,4,14,8,5             49.00               44.00  
2752     2020-03-16       0,1,0,0,2,0,0              3.00                2.00

In [27]:
# Now the code below helped split the numbers in the brackets.
w['visits_by_day'].str.split(',')

466             [11, 5, 1, 2, 1, 7, 7]
552         [13, 13, 13, 5, 13, 2, 11]
1399            [10, 4, 4, 6, 4, 0, 0]
2694            [5, 8, 5, 4, 14, 8, 5]
2752             [0, 1, 0, 0, 2, 0, 0]
                       ...            
14830911         [0, 0, 0, 0, 1, 1, 1]
14830972         [0, 0, 1, 0, 0, 0, 0]
14831166         [0, 1, 1, 0, 0, 0, 0]
14831510         [2, 4, 4, 1, 1, 1, 0]
14831677         [7, 6, 6, 4, 3, 8, 4]
Name: visits_by_day, Length: 34328, dtype: object

In [28]:
# Now, the columns are named in such a way that it will be easier to reshape the data later.
# 13 variables are now in the 'wide sample.'
w[['daily_visits1', 'daily_visits2', 'daily_visits3', 'daily_visits4','daily_visits5', 'daily_visits6', 'daily_visits7']] = w['visits_by_day'].str.split(',', expand = True)

In [29]:
# The 5th column contains the same information as the new columns made in the above code.
# The 5th column will be dropped now.
w.drop(columns = ['visits_by_day'])

safegraph_place_id region date_range_start  \
466       sg:0e5816c2c3074793ae5fc6f9c3e6bb7e     WY       2020-03-09   
552       sg:111ac71acaee4d10b00ef01388ae3e98     WY       2020-03-09   
1399      sg:2a17146a35e149dabce27332a52f8d1f     WY       2020-03-09   
2694      sg:5380bbb5685d4e6eb8427f4b1d418772     WY       2020-03-09   
2752      sg:55dbc9ae0c514561b30d27f7b7d74710     WY       2020-03-09   
...                                       ...    ...              ...   
14830911  sg:df3c053742584c8898672209b1a86995     WY       2020-03-30   
14830972  sg:e164b2b517294293a853af2e62c4cd25     WY       2020-03-30   
14831166  sg:e8436f667fcc48b8ac54889fee4e8f2b     WY       2020-03-30   
14831510  sg:f6114bc6d2634828b143f0abcff21dc6     WY       2020-03-30   
14831677  sg:fd3bfb53b0cc4ca7bfc051d1ce2bb2e5     WY       2020-03-30   

         date_range_end  raw_visit_counts  raw_visitor_counts daily_visits1  \
466          2020-03-16             34.00               19.00            11   
552          2020-03-16             70.00               41.00            13   
1399         2020-03-16             28.00               13.00            10   
2694         2020-03-16             49.00               44.00             5   
2752         2020-03-16              3.00                2.00             0   
...                 ...               ...                 ...           ...   
14830911     2020-04-06              3.00                3.00             0   
14830972     2020-04-06              1.00                1.00             0   
14831166     2020-04-06              2.00                2.00             0   
14831510     2020-04-06             13.00                8.00             2   
14831677     2020-04-06             38.00               35.00             7   

         daily_visits2 daily_visits3 daily_visits4 daily_visits5  \
466                  5             1             2             1   
552                 13            13             5            13   
1399                 4             4             6             4   
2694                 8             5             4            14   
2752                 1             0             0             2   
...                ...           ...           ...           ...   
14830911             0             0             0             1   
14830972             0             1             0             0   
14831166             1             1             0             0   
14831510             4             4             1             1   
14831677             6             6             4             3   

         daily_visits6 daily_visits7  
466                  7             7  
552                  2            11  
1399                 0             0  
2694                 8             5  
2752                 0             0  
...                ...           ...  
14830911             1             1  
14830972             0             0  
14831166             0             0  
14831510             1             0  
14831677             8             4  

[34328 rows x 13 columns]

The answer to question 2b is 13 variables are now in the 'wide sample.'

Moving on, question 2c. First, it mentions, reshape “wide sample” so that each observation is a place-date. Then after reshaping all the seven variables (dailyvisits1, …, dailyvisits7), consolidate the data into one variable (daily_visits). The reshaped data will be called “long sample”.

Question 2C ask, "how many observations do you have in the “long sample”? How many variables do you have in the “long sample”? Please comment on the pros and cons of the “long sample” relative to the “wide sample”."

In [30]:
# The "wide" data (w) was reshaped using the code below. It will be called “wlong”.
w['id'] = w.index
wlong = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'Rank')
wlong.reset_index(inplace = True)
# long = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'rank', sep = 'visits')
wlong.head()

id  Rank date_range_start date_range_end       visits_by_day region  \
0   466     1       2020-03-09     2020-03-16      11,5,1,2,1,7,7     WY   
1   552     1       2020-03-09     2020-03-16  13,13,13,5,13,2,11     WY   
2  1399     1       2020-03-09     2020-03-16      10,4,4,6,4,0,0     WY   
3  2694     1       2020-03-09     2020-03-16      5,8,5,4,14,8,5     WY   
4  2752     1       2020-03-09     2020-03-16       0,1,0,0,2,0,0     WY   

   raw_visitor_counts  raw_visit_counts                   safegraph_place_id  \
0               19.00             34.00  sg:0e5816c2c3074793ae5fc6f9c3e6bb7e   
1               41.00             70.00  sg:111ac71acaee4d10b00ef01388ae3e98   
2               13.00             28.00  sg:2a17146a35e149dabce27332a52f8d1f   
3               44.00             49.00  sg:5380bbb5685d4e6eb8427f4b1d418772   
4                2.00              3.00  sg:55dbc9ae0c514561b30d27f7b7d74710   

  daily_visits  
0           11  
1           13  
2           10  
3            5  
4            0

In [31]:
# In order to answer 2C, the code below is written and ran.
wlong.head()

id  Rank date_range_start date_range_end       visits_by_day region  \
0   466     1       2020-03-09     2020-03-16      11,5,1,2,1,7,7     WY   
1   552     1       2020-03-09     2020-03-16  13,13,13,5,13,2,11     WY   
2  1399     1       2020-03-09     2020-03-16      10,4,4,6,4,0,0     WY   
3  2694     1       2020-03-09     2020-03-16      5,8,5,4,14,8,5     WY   
4  2752     1       2020-03-09     2020-03-16       0,1,0,0,2,0,0     WY   

   raw_visitor_counts  raw_visit_counts                   safegraph_place_id  \
0               19.00             34.00  sg:0e5816c2c3074793ae5fc6f9c3e6bb7e   
1               41.00             70.00  sg:111ac71acaee4d10b00ef01388ae3e98   
2               13.00             28.00  sg:2a17146a35e149dabce27332a52f8d1f   
3               44.00             49.00  sg:5380bbb5685d4e6eb8427f4b1d418772   
4                2.00              3.00  sg:55dbc9ae0c514561b30d27f7b7d74710   

  daily_visits  
0           11  
1           13  
2           10  
3            5  
4            0

In [32]:
# Then the code below is also ran to answer 2C much easier.
wlong.shape

(240296, 10)

The anser to 2C is 240,296 observations and 8 variables are in the long data.

Here is a comment about the difference between 'wide' and 'long' sample. The data in wide sample have columns labeled 'daily_visit1,' 'daily_visit2,' etc. The data with such columns are considered as wide data since the data is expanded to show everything in the data. 

However, for example after stacking the columns with the label 'daily_visit1,' 'daily_visit2,' etc into only one or two columns, then running a command on one column (after it is stacked) it will change all the data. This change consolidates columns expanded. this is what is called a "long" data.

Moving on to question 2d. It mentions, "in your “long sample”, describe the variable dailyvisits. What are the mean, standard deviation, minimum, and maximum of this variable? Comment on the presence and extent of outliers. You can use whatever definition of “outlier”, as long as you state the definition clearly."

In [33]:
# This code shows what type 'daily_visits' is.
wlong.dtypes

id                             int64
Rank                           int64
date_range_start      datetime64[ns]
date_range_end        datetime64[ns]
visits_by_day                 object
region                        object
raw_visitor_counts           float64
raw_visit_counts             float64
safegraph_place_id            object
daily_visits                  object
dtype: object

Since 'daily_visits' was shown to be an object, it can not be described in the manner question 2d would like it to be. Before may answer question 2d, the variable 'daily_visits' needs to be changed into an interger. Python will not allow for an object to be described in the manner asked in question 2d.

In [34]:
# daily_visits is changed into a float in order to answer Q-2d using the command below.
wlong['daily_visits'] = wlong['daily_visits'].astype(float)

In [35]:
wlong['daily_visits'].describe()

count   240296.00
mean         3.81
std         11.06
min          0.00
25%          0.00
50%          1.00
75%          4.00
max        675.00
Name: daily_visits, dtype: float64

Here question 2d which mentions, "in your “long sample”, describe the variable daily_visits is answered.

**What are the mean? The mean is 3.81 visits daily** 

**What is the standard deviation? The standard deviation is 11.06.**

**What is the minimum? The minimum is 0.**

**What is the maximum of this variable? The maximum is 675 daily visit**

**Comment on the presence and extent of outliers. You can use whatever definition of “outlier”, as long as you state the definition clearly."

An outlier can be identified using two formulas. The formulas will find a value let's label it X. And if any daily_visits values is less than X when using formula (quartile one minus [1.5 times the interquartile range (IQR)]) then it is a lower outlier. If any daily_visits values is greater than X when using formula (quartile three plus (1.5*IQR)) it is to be considered an upper outlier.

Are there any ouliers? A lower outlier is not present. Numerous upper outliers are present. 675 is an upper outlier in this instance.

# Part 3: Data summary

Now it's time to summarize the “long sample.”

Now question 3a states, "tabulate calendar weeks in your long sample – that is, generate a table that reports the starting and ending dates of each week, # of unique weeks, and # of observations in each unique week."

In [36]:
daynames = ['Mon', 'Tue', 'Wed', 'Thur', 'Fri', 'Sat', 'Sun']

In [37]:
wlong['dayname'] = wlong['Rank'].replace([1,2,3,4,5,6,7], daynames)

In [38]:
wlong['dayname'] = pd.Categorical(wlong['dayname'], ordered = True, categories=daynames)

In [39]:
# This code tbaulates the calender weeks and reports the staring and ending dates.
# It also states the number of unique weeks (4). Finally, the number of observations in each unique week is displayed.
wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': ['first', 'size']})

date_range_end       
                          first   size
date_range_start                      
2020-03-09           2020-03-16  61537
2020-03-16           2020-03-23  59948
2020-03-23           2020-03-30  59409
2020-03-30           2020-04-06  59402

Relating to question 3a, the starting and ending dates of each week are displayed above. 
The number of unique weeks is 4. Finally the number of observations in each unique week is displayed.

Question 3b mentions to tabulate daily_visits by calendar dates in the long sample – that is, generate a table that tells you # of unique dates, # of observations per date, and the mean of dailyvisits per date.

In [40]:
# The code below and especially the third code below this one when ran will genrate a table
# that describes the number of unique date.
wlong['date'] = wlong['date_range_start'] + pd.to_timedelta(wlong['Rank'] - 1, unit = 'd') 

In [41]:
wlong[['date', 'daily_visits']].groupby('date').agg({'daily_visits':['size','mean']})

daily_visits     
                   size mean
date                        
2020-03-09         8791 5.69
2020-03-10         8791 5.66
2020-03-11         8791 5.80
2020-03-12         8791 5.82
2020-03-13         8791 6.00
2020-03-14         8791 4.32
2020-03-15         8791 3.45
2020-03-16         8564 4.50
2020-03-17         8564 4.43
2020-03-18         8564 4.26
2020-03-19         8564 3.58
2020-03-20         8564 3.05
2020-03-21         8564 2.10
2020-03-22         8564 2.47
2020-03-23         8487 3.74
2020-03-24         8487 3.61
2020-03-25         8487 3.95
2020-03-26         8487 3.39
2020-03-27         8487 3.26
2020-03-28         8487 2.86
2020-03-29         8487 2.41
2020-03-30         8486 3.45
2020-03-31         8486 3.41
2020-04-01         8486 3.30
2020-04-02         8486 3.00
2020-04-03         8486 3.50
2020-04-04         8486 2.92
2020-04-05         8486 2.28

In [42]:
# The number of observations per date and the mean of daily_visits per date is displayed when the code below is ran.
wlong.groupby('date').agg({'daily_visits': ['mean','size']})

daily_visits      
                   mean  size
date                         
2020-03-09         5.69  8791
2020-03-10         5.66  8791
2020-03-11         5.80  8791
2020-03-12         5.82  8791
2020-03-13         6.00  8791
2020-03-14         4.32  8791
2020-03-15         3.45  8791
2020-03-16         4.50  8564
2020-03-17         4.43  8564
2020-03-18         4.26  8564
2020-03-19         3.58  8564
2020-03-20         3.05  8564
2020-03-21         2.10  8564
2020-03-22         2.47  8564
2020-03-23         3.74  8487
2020-03-24         3.61  8487
2020-03-25         3.95  8487
2020-03-26         3.39  8487
2020-03-27         3.26  8487
2020-03-28         2.86  8487
2020-03-29         2.41  8487
2020-03-30         3.45  8486
2020-03-31         3.41  8486
2020-04-01         3.30  8486
2020-04-02         3.00  8486
2020-04-03         3.50  8486
2020-04-04         2.92  8486
2020-04-05         2.28  8486

In [43]:
# This code show the tbale of the long data before coding to answer question 3b.
wlong.head()

id  Rank date_range_start date_range_end       visits_by_day region  \
0   466     1       2020-03-09     2020-03-16      11,5,1,2,1,7,7     WY   
1   552     1       2020-03-09     2020-03-16  13,13,13,5,13,2,11     WY   
2  1399     1       2020-03-09     2020-03-16      10,4,4,6,4,0,0     WY   
3  2694     1       2020-03-09     2020-03-16      5,8,5,4,14,8,5     WY   
4  2752     1       2020-03-09     2020-03-16       0,1,0,0,2,0,0     WY   

   raw_visitor_counts  raw_visit_counts                   safegraph_place_id  \
0               19.00             34.00  sg:0e5816c2c3074793ae5fc6f9c3e6bb7e   
1               41.00             70.00  sg:111ac71acaee4d10b00ef01388ae3e98   
2               13.00             28.00  sg:2a17146a35e149dabce27332a52f8d1f   
3               44.00             49.00  sg:5380bbb5685d4e6eb8427f4b1d418772   
4                2.00              3.00  sg:55dbc9ae0c514561b30d27f7b7d74710   

   daily_visits dayname       date  
0         11.00     Mon 2020-03-09  
1         13.00     Mon 2020-03-09  
2         10.00     Mon 2020-03-09  
3          5.00     Mon 2020-03-09  
4          0.00     Mon 2020-03-09

In [44]:
wlong.tail(10)

id  Rank date_range_start date_range_end  visits_by_day region  \
240286  14830053     7       2020-03-30     2020-04-06  0,0,2,2,0,1,0     WY   
240287  14830090     7       2020-03-30     2020-04-06  2,5,1,3,6,1,4     WY   
240288  14830294     7       2020-03-30     2020-04-06  2,1,1,1,0,0,0     WY   
240289  14830883     7       2020-03-30     2020-04-06  0,0,1,0,1,2,0     WY   
240290  14830898     7       2020-03-30     2020-04-06  0,0,0,1,0,0,1     WY   
240291  14830911     7       2020-03-30     2020-04-06  0,0,0,0,1,1,1     WY   
240292  14830972     7       2020-03-30     2020-04-06  0,0,1,0,0,0,0     WY   
240293  14831166     7       2020-03-30     2020-04-06  0,1,1,0,0,0,0     WY   
240294  14831510     7       2020-03-30     2020-04-06  2,4,4,1,1,1,0     WY   
240295  14831677     7       2020-03-30     2020-04-06  7,6,6,4,3,8,4     WY   

        raw_visitor_counts  raw_visit_counts  \
240286                4.00              5.00   
240287               18.00             22.00   
240288                4.00              5.00   
240289                2.00              4.00   
240290                1.00              2.00   
240291                3.00              3.00   
240292                1.00              1.00   
240293                2.00              2.00   
240294                8.00             13.00   
240295               35.00             38.00   

                         safegraph_place_id  daily_visits dayname       date  
240286  sg:bef1c3b83bd94a97b0385f125c5acbf0          0.00     Sun 2020-04-05  
240287  sg:c03aab00128741e3a368e15d9bc991b1          4.00     Sun 2020-04-05  
240288  sg:c913a0766e484a7c96249acb3c18be4e          0.00     Sun 2020-04-05  
240289  sg:de35af72c7d7453e8bb3ac682692608b          0.00     Sun 2020-04-05  
240290  sg:deae2de64e53445cabce12b21ca34351          1.00     Sun 2020-04-05  
240291  sg:df3c053742584c8898672209b1a86995          1.00     Sun 2020-04-05  
240292  sg:e164b2b517294293a853af2e62c4cd25          0.00     Sun 2020-04-05  
240293  sg:e8436f667fcc48b8ac54889fee4e8f2b          0.00     Sun 2020-04-05  
240294  sg:f6114bc6d2634828b143f0abcff21dc6          0.00     Sun 2020-04-05  
240295  sg:fd3bfb53b0cc4ca7bfc051d1ce2bb2e5          4.00     Sun 2020-04-05

In [45]:
wlong['date'].unique()

array(['2020-03-09T00:00:00.000000000', '2020-03-16T00:00:00.000000000',
       '2020-03-23T00:00:00.000000000', '2020-03-30T00:00:00.000000000',
       '2020-03-10T00:00:00.000000000', '2020-03-17T00:00:00.000000000',
       '2020-03-24T00:00:00.000000000', '2020-03-31T00:00:00.000000000',
       '2020-03-11T00:00:00.000000000', '2020-03-18T00:00:00.000000000',
       '2020-03-25T00:00:00.000000000', '2020-04-01T00:00:00.000000000',
       '2020-03-12T00:00:00.000000000', '2020-03-19T00:00:00.000000000',
       '2020-03-26T00:00:00.000000000', '2020-04-02T00:00:00.000000000',
       '2020-03-13T00:00:00.000000000', '2020-03-20T00:00:00.000000000',
       '2020-03-27T00:00:00.000000000', '2020-04-03T00:00:00.000000000',
       '2020-03-14T00:00:00.000000000', '2020-03-21T00:00:00.000000000',
       '2020-03-28T00:00:00.000000000', '2020-04-04T00:00:00.000000000',
       '2020-03-15T00:00:00.000000000', '2020-03-22T00:00:00.000000000',
       '2020-03-29T00:00:00.000000000', '2020-04-05

In [46]:
# 28 unique dates
wlong['date'].nunique()

28

Relating to question 3b, the number of observations per date, and the mean of dailyvisits per date is displayed when the code below is ran. .

Question 3c requires one to define a new variable “dayofweek”. You can define the variable as numeric, with the value 1 for Monday, and …7 for Sunday; or as text with the values being “Monday”,..,”Sunday”.

 

Also, it requires one to tabulate dailyvisits by dayofweek in your long sample – that is, generate a table that tells you # of unique values in 

dayofweek, # of observations per value, and the mean of dailyvisits per value of dayofweek. On average, which day of week has 

the highest dailyvisits in your long sample?

In [47]:
# The code below shows a table where the column 'Rank' displays the 'dayofweek' as a numeric value.
# 1 is for Monday and 2 is for Tuesday, and so forth
wlong.head()

id  Rank date_range_start date_range_end       visits_by_day region  \
0   466     1       2020-03-09     2020-03-16      11,5,1,2,1,7,7     WY   
1   552     1       2020-03-09     2020-03-16  13,13,13,5,13,2,11     WY   
2  1399     1       2020-03-09     2020-03-16      10,4,4,6,4,0,0     WY   
3  2694     1       2020-03-09     2020-03-16      5,8,5,4,14,8,5     WY   
4  2752     1       2020-03-09     2020-03-16       0,1,0,0,2,0,0     WY   

   raw_visitor_counts  raw_visit_counts                   safegraph_place_id  \
0               19.00             34.00  sg:0e5816c2c3074793ae5fc6f9c3e6bb7e   
1               41.00             70.00  sg:111ac71acaee4d10b00ef01388ae3e98   
2               13.00             28.00  sg:2a17146a35e149dabce27332a52f8d1f   
3               44.00             49.00  sg:5380bbb5685d4e6eb8427f4b1d418772   
4                2.00              3.00  sg:55dbc9ae0c514561b30d27f7b7d74710   

   daily_visits dayname       date  
0         11.00     Mon 2020-03-09  
1         13.00     Mon 2020-03-09  
2         10.00     Mon 2020-03-09  
3          5.00     Mon 2020-03-09  
4          0.00     Mon 2020-03-09

In [48]:
wlong[["Rank"]].groupby("Rank").agg({'Rank': ['first', 'size']})

Rank       
     first   size
Rank             
1        1  34328
2        2  34328
3        3  34328
4        4  34328
5        5  34328
6        6  34328
7        7  34328

Question 3c (1 point):  Define a new variable “dayofweek”. You can define the variable as numeric, with the value 1 for Monday, 

and …7 for Sunday; or as text with the values being “Monday”,..,”Sunday”.

 

Tabulate dailyvisits by dayofweek in your long sample – that is, generate a table that tells you # of unique values in 

dayofweek, # of observations per value, and the mean of dailyvisits per value of dayofweek. On average, which day of week has 

the highest dailyvisits in your long sample?

Question 3d (1 point): Define a new variable “positivevisits” equal to 1 if dailyvisits>0 and 0 otherwise. Tabulate dayofweek 

and positivevisits – that is, generate a table that tells you # of observations in each possible combination of dayofweek and 

positivevisits. How many observations have positivevisits=0 on Saturday? How many observations have positivevisits=1 on 

Wednesday?

In [49]:
# 3D This line creates the positive_visits variable
wlong['positive_visits'] = np.where(wlong['daily_visits'] > 0, 1, 0)

In [50]:
# The code below creates a table that generates a table that displays the number of observations in each posible combination of dayofweek.
wlong.groupby(['dayname', 'positive_visits']).size().unstack()

positive_visits      0      1
dayname                      
Mon               9417  24911
Tue               9424  24904
Wed               9427  24901
Thur             10413  23915
Fri              10805  23523
Sat              14943  19385
Sun              16662  17666

Question 3e (2 points): Define a new variable “weekend” equal to 1 if dayofweek is Saturday or Sunday, and 0 otherwise.

 

Tabulate weekend – that is, generate a table that tells you # of observations in weekend days vs non-weekend days. 

 

Tabulate dailyvisits by weekend – that is, generate a table that reports the mean of dailyvisits on a weekend day vs a non-weekend day. On average, do you observe more or fewer dailyvisits on a weekend day? Could you name at least two reasons that may explain your finding?

In [51]:
wlong.head(1000)

id  Rank date_range_start date_range_end        visits_by_day region  \
0       466     1       2020-03-09     2020-03-16       11,5,1,2,1,7,7     WY   
1       552     1       2020-03-09     2020-03-16   13,13,13,5,13,2,11     WY   
2      1399     1       2020-03-09     2020-03-16       10,4,4,6,4,0,0     WY   
3      2694     1       2020-03-09     2020-03-16       5,8,5,4,14,8,5     WY   
4      2752     1       2020-03-09     2020-03-16        0,1,0,0,2,0,0     WY   
..      ...   ...              ...            ...                  ...    ...   
995  460427     1       2020-03-09     2020-03-16        2,1,0,2,1,0,0     WY   
996  461427     1       2020-03-09     2020-03-16        0,2,1,2,5,0,1     WY   
997  461618     1       2020-03-09     2020-03-16        1,4,0,0,0,0,0     WY   
998  462830     1       2020-03-09     2020-03-16  10,7,13,14,12,23,11     WY   
999  463021     1       2020-03-09     2020-03-16        7,2,0,3,0,2,7     WY   

     raw_visitor_counts  raw_visit_counts  \
0                 19.00             34.00   
1                 41.00             70.00   
2                 13.00             28.00   
3                 44.00             49.00   
4                  2.00              3.00   
..                  ...               ...   
995                5.00              6.00   
996               11.00             11.00   
997                2.00              5.00   
998               76.00             90.00   
999               11.00             21.00   

                      safegraph_place_id  daily_visits dayname       date  \
0    sg:0e5816c2c3074793ae5fc6f9c3e6bb7e         11.00     Mon 2020-03-09   
1    sg:111ac71acaee4d10b00ef01388ae3e98         13.00     Mon 2020-03-09   
2    sg:2a17146a35e149dabce27332a52f8d1f         10.00     Mon 2020-03-09   
3    sg:5380bbb5685d4e6eb8427f4b1d418772          5.00     Mon 2020-03-09   
4    sg:55dbc9ae0c514561b30d27f7b7d74710          0.00     Mon 2020-03-09   
..                                   ...           ...     ...        ...   
995  sg:2abf03c6d9dc445f9d4eefad0ce28fa3          2.00     Mon 2020-03-09   
996  sg:4ae24483c8a248f8a881466f05b2da58          0.00     Mon 2020-03-09   
997  sg:5105e746c85f49338b88eb8122c88bad          1.00     Mon 2020-03-09   
998  sg:7944c7bcb37b4faaa8a154fd38415fb5         10.00     Mon 2020-03-09   
999  sg:7f796a7f217b4af8a1074f2b35977385          7.00     Mon 2020-03-09   

     positive_visits  
0                  1  
1                  1  
2                  1  
3                  1  
4                  0  
..               ...  
995                1  
996                0  
997                1  
998                1  
999                1  

[1000 rows x 13 columns]

In [52]:
# This line of code creates the weekend variable. (I am missing a symbol that looks like the uppercase (I) between the parenthesis after 6 and before (wlong.)
wlong['weekend'] = np.where(wlong['dayname'].isin(['Sat', 'Sun']), 1, 0)

In [53]:
wlong['weekend'].value_counts()

0    171640
1     68656
Name: weekend, dtype: int64

In [54]:
wlong.groupby('weekend').agg({'daily_visits':'mean'})

daily_visits
weekend              
0                4.18
1                2.86

In [55]:
wlong.to_csv('wlong_raw.csv', index = False)

# Everything below is codes not used and were for reference.

In [56]:
# 3b 

import pandas as pd
#wlong.date_range_start.apply(lambda dt: dt.date()).groupby([wlong.daily_visits]).nunique()
#wlong.pd.to_datetime(df['date_range_']).dt.date.unique()


In [57]:
# 2C The reshaped data will be called “wlong_sample”.
#w['id'] = w.index
#wlong = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'Rank')
#wlong.reset_index(inplace = True)
# long = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'rank', sep = 'visits')

In [58]:
# I'm trying to create a table of dailyvisits
# wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_start': 'second'})
#wlong = pd.wide_to_long(w, ['DailyVisits'], i = 'id', j = 'DailyVisits')

In [59]:
#daily_visits_by_calenderDate.head()

In [60]:
# Thiss command shows the five
#w.head()

In [61]:
#w['data_range_start'].unique()

In [62]:
# idk what this is
#w['id'] = w.index

In [63]:
#wlong_sample['DailyVisits']

In [64]:
#wlong['daily_visits'].nunique(6)

In [65]:
#wlong.describe()

In [66]:
#wlongwlong['date_range_end'].unique()

In [67]:
#wlong_sample['daily_visits'].describe()

In [68]:
#wlong.dtypes

In [69]:
# WY.to_csv(‘WY.csv’) This is extra

In [70]:
# wlong['dailly_visits'] = np.where(wlong['daily_visits'] == 'daily_visits', 'int')

In [71]:
# 57 is reshaping the data
# wlong = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'DayofWeek')

In [72]:
# 2C The reshaped data will be called “daily_visits” (long sample).
#w['id'] = w.index
#wlong = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'rank')
# long = pd.wide_to_long(w, ['daily_visits'], i = 'id', j = 'rank', sep = 'visits')

In [73]:
#wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})


#wlong['calender_week'] = np.where['date_range_start']

# wlong['dayofweek'] = wlong['date_range_start'].dt.dayofweek 
# .dt.day_name() simply gives you the day name.
# wlong['day_name'] = wlong['date_range_start'].dt.day_name()

# wlong[['date_range_start', 'dayofweek', 'day_name']]

In [74]:
#dayof_wk = wlong[wlong['day_name'] == 'monday']
#dayOf_wk

In [75]:
# rearrange column orders so that brings 'date', 'dayofweek', 'day_name' up front
#first_cols = ['date_range_start', 'dayofweek', 'day_name','date_range_end', 'dayofweek', 'day_name']
#other_cols = [col for col in wlong.columns if col not in first_cols]
#wlong = wlong[first_cols + other_cols]
#wlong.head()

In [76]:
#wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})
#wlong['calender_week'] = wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})

#wlong['calender_week'] = np.where['date_range_start']

#wlong['dayofweek'] = wlong['daily_visits'].dt.dayofweek 
# .dt.day_name() simply gives you the day name.
#wlong['day_name'] = wlong[daily_visit].dt.day_name()

#wlong[['daily_visits', 'dayofweek', 'day_name']]

In [77]:
# to get the day of week: (remember python index starts from 0!)
# .dt.dayofweek transforms the date into day of week, in numbers, with 0 being Monday.
#data2['dayofweek'] = data2['date'].dt.dayofweek 
# .dt.day_name() simply gives you the day name.
# data2['day_name'] = data2['date'].dt.day_name()

# data2[['date', 'dayofweek', 'day_name']]

In [78]:
#wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end':'first'})

# wlong.groupby(['date_range_start']).agg({'date_range_end': ['size', 'first'], 'date_range_start': 'first'})

#wlong.groupby(['date_range_start']).agg({'date_range_end': ['size', 'first'], 'date_range_start': 'first'})


In [79]:
#wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})
#wlong['calender_week'] = wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})

#wlong['calender_week'] = np.where['date_range_start']

#wlong['dayofweekend'] = wlong['date_range_end'].dt.dayofweek 
# .dt.day_name() simply gives you the day name.
#wlong['day_name'] = wlong['date_range_end'].dt.day_name()

#wlong[['date_range_end', 'dayofweek', 'day_name']]

In [80]:
# compare = wlong_sample[['date_range_start', 'date_range_end']].groupby('DailyVisits').agg('daily_visits':'sum',
#                                                                               'visit_by_day'
#:'raw_visitor_counts'))

In [81]:
#wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})

#wlong['calender_week'] = wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': 'first'})

#wlong['calender_week'] = np.where['date_range_start']

In [82]:
# wlong[['date_range_start', 'date_range_end']].groupby('date_range_start').agg({'date_range_end': ['first', 'size']})


In [83]:
#wlong['visits_by_day'] = wlong['visits_by_day']

In [84]:
# daily_visits_by_calenderDate = wlong[wlong['day_name'] == 'Monday']
# daily_visits_by_calenderDate

# Each observation is based on different time period.
# We want to standardize them so that we can compare across observations and across apps.
# We transfer all data usage into 30-day period.

#wlong['daily_visits'] = wlong['daily_visits']/wlong["visits_by_day"] * 7
#wlong.head(3)

In [85]:
# wlong['daily_visits_by_calenderDate'] = pd.read. w['date_range_start'].unique()